_articles/incremental-deploys.md

```diff
@@ -0,0 +1,215 @@
+infrastructure update—opens up new tools for protecting users from bad releases.
+
+We’ll begin with the simple routes we set up previously
+[on your laptop](on-your-laptop.html).
+Next, we’ll extend that config to release a new version of one of the services
+using traffic shifting. We’ll also cover header-based routing and weighted load
+balancing to show how to use traffic management to canary a release, first to
+special requests (e.g. requests from your laptop), then to a small fraction of
+all requests.
+
+## The setup
+
+For this guide, we’ll need:
+
+- [Docker](https://docs.docker.com/install/)
+- [Docker Compose](https://docs.docker.com/compose/install/)
+- [Git](https://help.github.com/articles/set-up-git/)
+- [curl](https://curl.haxx.se/)
+
+## Header-based Routing
+
+First, we'll create a new version of service1 to illustrate the value of
+header-based routing for our services. Still following along with the
+[front-proxy](https://github.com/envoyproxy/envoy/tree/master/examples/front-proxy)
+example from the Envoy repo, modify the
+[docker-compose.yml](https://github.com/envoyproxy/envoy/blob/master/examples/front-proxy/docker-compose.yml)
+to spin up a new service, called `service1a`.
+
+```yaml
+  service1a:
+    build:
+      context: .
+      dockerfile: Dockerfile-service
+    volumes:
+      - ./service-envoy.yaml:/etc/service-envoy.yaml
+    networks:
+      envoymesh:
+        aliases:
+          - service1a
+    environment:
+      - SERVICE_NAME=1a
+    expose:
+      - "80"
+```
+
+To make sure Envoy can discover this service, we’ll also add it to the
+`clusters’ section of our configuration file.
+
+```yaml
+  - name: service1a
+    connect_timeout: 0.25s
+    type: strict_dns
+    lb_policy: round_robin
+    http2_protocol_options: {}
+    hosts:
+    - socket_address:
+        address: service1a
+        port_value: 80
+```
+
+To make this routable, we can add a new route with a `headers` field in the
+match. Since routing rules are applied in order, we’ll add this rule to the top
+of the `routes` key. Requests with a header that matches our new rule will be
+sent to our new service, while requests that don't include this header will
+still get service 1.
+
+```yaml
+routes:
+- match:
+    prefix: "/service/1"
+    headers:
+      - name: "x-canary-version"
+        value: "service1a"
+  route:
+    cluster: service1a
+- match:
+    prefix: "/service/1"
+  route:
+    cluster: service1
+- match:
+    prefix: "/service/2"
+  route:
+    cluster: service2
+```
+
+Shut down and then relaunch our example services with:
+
+`docker-compose down --remove-orphans`
+
+`docker-compose up --build -d`
+
+In a production Envoy deployment, configuration changes like this won’t require
+a restart of Envoy, but since we’re running everything locally, we aren’t able
+to take advantage of its dynamic configuration abilities.
+
+If we make a request to our service with no headers, we'll get a response
+from service 1:
+
+```console
```

Comment:
for lines with commands and then output, you'll get a copyable command if you do it like this (note the `$` vs `>`):

```console
$ curl localhost:8000/service/1
```

```shell
Hello from behind Envoy (service 1)! hostname: d0adee810fc4 resolvedhostname: 172.18.0.2
```